### PR TITLE
Lpfg 940 Create booking db tables and classes

### PR DESCRIPTION
**Pull request by Peter-D-Taylor**
Created V1.1.5__add-event-learner-booking-tables.sql in h2 and mysql

Created domain classes:
Booking.java
BookingIdentity.java
Event.java
Learner.java

Created repository classes:
BookingRepository.java
EventRepository.java
LeanerRepository.java

### CODE REVIEW

File: src/main/java/uk/gov/cslearning/record/domain/Booking.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import org.codehaus.jackson.annotate.JsonIgnore;
+
+import javax.persistence.*;
+import java.time.LocalDateTime;
+
+@Entity
+public class Booking {
+
+    @JsonIgnore
+    @EmbeddedId
```
Comment (robleney-kainos) [inline src/main/java/uk/gov/cslearning/record/domain/Booking.java]: Can you make the primary key an auto increment long please?

File: src/main/java/uk/gov/cslearning/record/domain/Event.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package uk.gov.cslearning.record.domain;
+
+import javax.persistence.*;
+
+import static com.google.gson.internal.$Gson$Preconditions.checkArgument;
```
Comment (robleney-kainos) [inline src/main/java/uk/gov/cslearning/record/domain/Event.java]: Is this needed?

File: src/main/java/uk/gov/cslearning/record/domain/Booking.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import com.fasterxml.jackson.datatype.jsr310.ser.LocalDateTimeSerializer;
+
+import javax.persistence.*;
+import java.time.LocalDateTime;
+
+@Entity
+public class Booking {
+
+    @Id
+    private Long id;
```
Comment (robleney-kainos) [inline src/main/java/uk/gov/cslearning/record/domain/Booking.java]: Sorry this also needs to be an auto increment.

File: src/main/resources/db/migration/h2/V1.1.5__add-event-learner-booking-tables.sql [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  `learner_id` bigint NOT NULL,
+  `event_id` bigint NOT NULL,
+  `payment_details` varchar(255),
+  `status` varchar(9) NOT NULL,
+  `booking_time` datetime NOT NULL,
+  CONSTRAINT `status_type` CHECK (`status` IN ('Requested', 'Confirmed')),
+  CONSTRAINT `FK_booking_learnerId_learner_id` FOREIGN KEY(`learner_id`) REFERENCES `learner`(`id`) ON DELETE CASCADE ON UPDATE CASCADE,
+  CONSTRAINT `FK_booking_eventId_event_id` FOREIGN KEY(`event_id`) REFERENCES  `event`(`id`) ON DELETE CASCADE ON UPDATE CASCADE,
+  PRIMARY KEY(`id`)
+);'' ||
```
Comment (elxsy) [inline src/main/resources/db/migration/h2/V1.1.5__add-event-learner-booking-tables.sql]: `||` ? typo?

File: src/main/resources/db/migration/mysql/V1.1.5__add-event-learner-booking-tables.sql [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+);
+
+CREATE TABLE `learner` (
+  `id` bigint NOT NULL AUTO_INCREMENT,
+  `uuid` varchar(60) NOT NULL,
+  PRIMARY KEY(`id`)
+);
+
+CREATE TABLE `booking` (
+  `id` bigint NOT NULL AUTO_INCREMENT,
```
Comment (elxsy) [inline src/main/resources/db/migration/mysql/V1.1.5__add-event-learner-booking-tables.sql]: `unsigned int` should be more than enough for all the integers fields

File: src/main/resources/db/migration/mysql/V1.1.5__add-event-learner-booking-tables.sql [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  `learner_id` bigint NOT NULL,
+  `event_id` bigint NOT NULL,
+  `payment_details` varchar(255),
+  `status` varchar(9) NOT NULL,
+  `booking_time` datetime NOT NULL,
```
Comment (elxsy) [inline src/main/resources/db/migration/mysql/V1.1.5__add-event-learner-booking-tables.sql]: can we order them in this state please?
```
booking_time,
status,
payment_details
```

Status_type check shouldn't be ENUM and all UPPERCASE?

File: src/main/resources/db/migration/h2/V1.1.5__add-event-learner-booking-tables.sql [hunk 7]
```diff
@@ -0,0 +1,23 @@
+CREATE TABLE `event` (
+  `id` int NOT NULL AUTO_INCREMENT,
+  `path` varchar(255) NOT NULL,
+  PRIMARY KEY(`id`)
+);
+
+CREATE TABLE `learner` (
+  `id` int NOT NULL AUTO_INCREMENT,
+  `uuid` varchar(60) NOT NULL,
```
Comment (elxsy) [inline src/main/resources/db/migration/h2/V1.1.5__add-event-learner-booking-tables.sql]: naming suggestion `uid`

File: src/main/java/uk/gov/cslearning/record/domain/Booking.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    private int eventId;
+
+    @Column
+    private String paymentDetails;
+
+    @Column(nullable = false)
+    private String status;
+
+    @JsonSerialize(using = LocalDateTimeSerializer.class)
+    private LocalDateTime bookingTime;
```
Comment (elxsy) [inline src/main/java/uk/gov/cslearning/record/domain/Booking.java]: Jakub started using `Instant` for this. Can you have a look into what are the pros vs cons of this?

**Pull request closed without merging**